module/zfs/zfs_ioctl.c

```diff
@@ -3646,7 +3646,11 @@ zfs_ioc_rollback(const char *fsname, nvlist_t *args, nvlist_t *outnvl)
 		}
 		deactivate_super(zsb->z_sb);
 	} else {
+		int e;
+		e = zvol_suspend(fsname);
 		error = dsl_dataset_rollback(fsname, NULL, outnvl);
+		if (e == 0)
+			zvol_resume(fsname);
```

Comment:
It would be nice if we could use the same basic construct as above.  You're almost there with this patch already.  It could be done by adding a helper function similar to `get_zfs_sb()`.  The existing `zvol_find_by_name()` gets us almost all the way there, a `zvol_state_t` would serve the same purpose as a `zfs_sb_t`.  Then a proper owner could be  passed to `dsl_dataset_rollback()` making the zvol case basically the same as the owned filesystem case.